utils/i18n.tsx

```diff
@@ -0,0 +1,211 @@
+    ns: config.namespaces, // String or array of namespaces to load
+    defaultNS: config.defaultNamespace, // Default namespace used if not passed to translation function
+    interpolation: {
+      escapeValue: false, // Not needed for react as it escapes by default
+    },
+    react: {
+      useSuspense: false, // Not compatible with SSR
+    },
+    load: 'languageOnly', // Remove if you want to use localization (en-US, en-GB)
+  });
+
+  return i18next;
+};
+
+let globalI18nextInstance: any = null;
+
+/**
+ * Returns a singleton instance of the i18next object with the specified language and locales.
+ * If the instance doesn't exist, it creates a new one; otherwise, it changes the language of the existing instance.
+ * @param {string} language - The language code representing the desired language.
+ * @param {Object} locales - The locales object containing translations for different languages and namespaces.
+ * @returns {Object} - The i18next instance.
+ */
+export const i18nextInstance = (language: string, locales: object): i18n => {
+  if (!globalI18nextInstance) {
+    globalI18nextInstance = createI18nextInstance(locales, language);
+
+    return globalI18nextInstance;
+  } else {
+    globalI18nextInstance.changeLanguage(language);
+
+    return globalI18nextInstance;
+  }
+};
+
+// Prevent rerender
+let loaded = false;
+
+/**
+ * A React component that provides i18n functionality to the child components.
+ * @param {Object} props - The props object containing the i18n options and child components.
+ * @returns {JSX.Element|null} - The child components wrapped in the i18n provider, or null if hydration is not allowed.
+ */
+export const I18nProvider = (props: any) => {
+  const [hydration, setHydration] = useState(false);
+
+  const options: any = { ...config, ...props.i18n };
+
+  if (!props.i18n?.locales) {
+    throw new Error('locales object was not passed into I18nProvider');
+  }
+
+  const router = useRouter();
+  const { asPath, query } = router;
+
+  // Detect the current language
+  const slug = asPath.split('/')[1];
+  const langSlug = config.languages.includes(slug) && slug;
+  const language = (
+    query.lang ||
+    langSlug ||
+    config.defaultLanguage
+  ).toString();
+
+  // Load only once, otherwise there will be an re-render infinite loop
+  if (!loaded) {
+    i18nextInstance(language, props.i18n.locales);
+  }
+
+  const { i18n } = useTranslation();
+
+  useEffect(() => {
+    // Overwrite the current store
+    i18n.services.resourceStore.data = props.i18n.locales;
+
+    // Required to display the updated translations
+    i18n.changeLanguage(language);
+  }, [i18n, language, props.i18n.locales]);
+
+  useEffect(() => {
+    loaded = true;
+    i18n.changeLanguage(language);
+  }, [i18n, language]);
+
+  useEffect(() => {
+    const hasWindow = typeof window !== 'undefined';
+    if (hasWindow && options.allowHydration) {
+      setHydration(true);
+    }
+  }, [options.allowHydration]);
```

Comment:
Kindly add the comments on why these `useEffects` block are used and what they are doing.